### PR TITLE
Nodal

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Push nodal E and B fields.  It's done in C++.

### CODE REVIEW

**Comment by RemiLehe** [comment]
Thanks! Could you check why tests are failing?

**Comment by WeiqunZhang** [comment]
Is there a way to rerun the travis tests?  It seems fine on my computer.  There were some changes unrelated to this pull request in AMReX after the pull request was made.   I updated warpx dev and nodal branches.   I wonder if the travis build used mismatched versions.

**Comment by WeiqunZhang** [comment]
I saw the log, but I could not find more details about the amrex version used in the build.  I am running the regression tests on he "nodal" branch on battra.  All tests finished so far have passed.

**Comment by WeiqunZhang** [comment]
What I mean by "mismatched version" is I pushed commits to both warpx and amrex github repos and travis might have used the new warpx but old amrex.

**Comment by RemiLehe** [comment]
OK, I just restarted the Travis tests and it seems that everything is working fine.
I guess this was indeed a case of mismatched version.

File: Source/WarpX_K.H [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#ifndef WARPX_K_H_
+#define WARPX_K_H_
+
+#include <AMReX_FArrayBox.H>
+
+using namespace amrex;
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void warpx_push_bx_nodal (int j, int k, int l, Array4<Real>& Bx,
+                          Array4<Real> const& Ey, Array4<Real> const& Ez, Real dtsdy, Real dtsdz)
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/WarpX_K.H]: I simply picked K for kernels.

Yes `Array4` has bound checking in debug mode.  The pointer inside is not owned by the object, so it does not deal with allocating and deallocating memory.  We use it to wrap the data in `FArrayBox`.

File: Source/WarpXEvolve.cpp [hunk 2]
```diff
@@ -471,7 +472,7 @@ WarpX::EvolveB (int lev, PatchType patch_type, amrex::Real dt)
 {
     const int patch_level = (patch_type == PatchType::fine) ? lev : lev-1;
     const std::array<Real,3>& dx = WarpX::CellSize(patch_level);
-    const std::array<Real,3> dtsdx {dt/dx[0], dt/dx[1], dt/dx[2]};
+    Real dtsdx = dt/dx[0], dtsdy = dt/dx[1], dtsdz = dt/dx[2];
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/WarpXEvolve.cpp]: I already learned it and I think it's fine.

File: Source/WarpXEvolve.cpp [hunk 3]
```diff
@@ -519,17 +537,18 @@ WarpX::EvolveB (int lev, PatchType patch_type, amrex::Real dt)
...
+        }
 
         if (cost) {
             Box cbx = mfi.tilebox(IntVect{AMREX_D_DECL(0,0,0)});
             if (patch_type == PatchType::coarse) cbx.refine(rr);
-            wt = (amrex::second() - wt) / cbx.d_numPts();\
-            FArrayBox* costfab = cost->fabPtr(mfi);
-            AMREX_LAUNCH_HOST_DEVICE_LAMBDA ( cbx, work_box,
+            wt = (amrex::second() - wt) / cbx.d_numPts();
+            auto costfab = cost->array(mfi);
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/WarpXEvolve.cpp]: `array(mfi)` returns `amrex::Array4<amrex::Real>`.  Using auto, one doesn't have to type that many characters.  The returned object is the same for cpu and gpu.

**Comment by MaxThevenet** [comment]
Thanks for your replies, I approved the PR.

File: Source/WarpXEvolve.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlvay) [reply Source/WarpXEvolve.cpp]: My bad for the French. This was propagated from parts that I developed for a code from a team at Ecole Polytechnique in France in early 2000s. I crosschecked that in the first code that I developed earlier, I used dtodx. Would be easy to switch. :-)

**Comment by RemiLehe** [comment]
@WeiqunZhang I added automated tests and deactivated `plot_divb` by default (it does not work for nodal, and it is not very useful in general). I'll merge as soon as the tests pass, so that we can then focus on the #56

**Pull request merged**